File: app/javascript/channels/ai_assistant_channel.js
import consumer from "./consumer"

document.addEventListener('DOMContentLoaded', function() {
  const journalEntryElement = document.querySelector('[data-entry-id]');
  
  if (journalEntryElement) {
    const entryId = journalEntryElement.getAttribute('data-entry-id');
    
    consumer.subscriptions.create({ channel: "AiAssistantChannel", entry_id: entryId }, {
      connected() {
        // Called when the subscription is ready for use on the server
        console.log('Connected to AI Assistant channel for entry', entryId);
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      received(data) {
        // Called when there's incoming data on the websocket for this channel
        const aiResponseDiv = document.getElementById('ai-response');
        
        if (data.status === 'success' && data.prompt) {
          aiResponseDiv.innerHTML = `<p class="italic text-gray-600">${data.prompt}</p>`;
        } else if (data.status === 'error') {
          aiResponseDiv.innerHTML = `<p class="italic text-red-600">Error: ${data.message}</p>`;
        }
      },
      
      requestPrompt(content) {
        this.perform('request_prompt', { content: content, entry_id: entryId });
      }
    });
  }
});

File: app/channels/ai_assistant_channel.rb
class AiAssistantChannel < ApplicationCable::Channel
  def subscribed
    stream_from "ai_assistant_channel_#{params[:entry_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def request_prompt(data)
    content = data["content"]
    entry_id = data["entry_id"]

    # Generate AI prompt asynchronously
    Thread.new do
      begin
        prompt = OllamaService.new.generate_prompt(content)

        # Broadcast the prompt back to the client
        ActionCable.server.broadcast(
          "ai_assistant_channel_#{entry_id}",
          { prompt: prompt, status: "success" }
        )
      rescue => e
        ActionCable.server.broadcast(
          "ai_assistant_channel_#{entry_id}",
          { status: "error", message: e.message }
        )
      end
    end
  end
end
